src/mds/SessionMap.cc

```diff
@@ -1040,12 +1040,35 @@ int Session::check_access(CInode *in, unsigned mask,
 			  const vector<uint64_t> *caller_gid_list,
 			  int new_uid, int new_gid)
 {
+  dout(20) << __func__ << ": " << *in
+           << " caller_uid=" << caller_uid
+           << " caller_gid=" << caller_gid
+           << " caller_gid_list=" << *caller_gid_list
+           << dendl;
+
   string path;
-  CInode *diri = NULL;
-  if (!in->is_base())
-    diri = in->get_projected_parent_dn()->get_dir()->get_inode();
-  if (diri && diri->is_stray()){
-    path = in->get_projected_inode()->stray_prior_path;
+  if (!in->is_base()) {
+    auto* dn = in->get_projected_parent_dn();
+    auto* pdiri = dn->get_dir()->get_inode();
+    if (pdiri) {
+      if (pdiri->is_stray()) {
+        path = in->get_projected_inode()->stray_prior_path;
+      } else if (!pdiri->is_base()) {
+        /* is the pdiri in the stray (is this inode in a snapshotted deleted directory?) */
+        auto* gpdiri = pdiri->get_projected_parent_dn()->get_dir()->get_inode();
+        /* stray_prior_path will not necessarily be part of the inode because
+         * it's set on unlink but that happens after the snapshot, naturally.
+         * We need to construct it manually.
+         */
+        if (gpdiri->is_stray()) {
+          /* just check access on the parent dir */
+          path = pdiri->get_projected_inode()->stray_prior_path;
```

Comment:
I don't think I quite understand the claim
> Files in that directory will not have correct stray_prior_path set because the snapshot was naturally taken before the file is unlinked.

We still have to unlink the file; why doesn't that set the stray_prior_path correctly?

Just as importantly, are we sure we're correct to grab permissions by iterating up directories like this? Can't we check_access on a directory inode and then go up to its parent and get permission denied, even if had access to the actual directory we were trying to operate on?

Reply:
> I don't think I quite understand the claim
> 
> > Files in that directory will not have correct stray_prior_path set because the snapshot was naturally taken before the file is unlinked.
> 
> We still have to unlink the file; why doesn't that set the stray_prior_path correctly?

`stray_prior_path` is part of the `inode_t` which is projected and set when the inode is unlinked. A snapshot of that inode would not have visibility of that since it's in a future project of the inode. (And we can't use it even if we force looking at it since the inode may have been renamed elsewhere since the snapshot.)

> Just as importantly, are we sure we're correct to grab permissions by iterating up directories like this?

Well, we need the `stray_prior_path` so that's what we have to do. But, it's not completely correct and maybe can't be with existing information, consider:

```
mkdir a
touch a/file
mkdir .snap/one
mv a b
rm -rf b # a's stray_prior_path is now "/b"
ls .snap/one/a/file # uses stray_prior_path "/b" but should be "/a"
```

Tracking the `stray_prior_path` correctly in this situation is difficult since we'd need to record the path changes during rename. I don't think it's worth the trouble when we're talking about access control on a read-only snapshot.

> Can't we check_access on a directory inode and then go up to its parent and get permission denied, even if had access to the actual directory we were trying to operate on?

No. Because this change only affects access control to non-directories. An unlinked directory inode's parent is always a stray directory. That is handled on line 1055 above.